### ISSUE TITLE
As a user I would like to Highlight my SQL code syntax in JupyterLab

### ISSUE THREAD

**Issue by julioyildo** [jupyterlab/jupyterlab#8879]
Hi,

**I am looking for a way to highlight SQL codes in JupyterLab.**

I was able to highlight SQL in the Jupyter Notebook with the classic Launcher  (Help > Launch Classic Notebook)  with the following code in the `.jupyter/custom/custom.js` file.

```
require(['notebook/js/codecell'], function (codecell) {
    codecell.CodeCell.options_default.highlight_modes['magic_text/x-mssql'] = { 'reg': [/%?%sql/] };
    Jupyter.notebook.events.one('kernel_ready.Kernel', function () {
        Jupyter.notebook.get_cells().map(function (cell) {
            if (cell.cell_type == 'code') { cell.auto_highlight(); }
        });
    });
});
```

![image](https://user-images.githubusercontent.com/17247165/90888366-6702a100-e3b6-11ea-8e55-cf097c2ac44a.png)
([credit image](https://stackoverflow.com/questions/57966403/how-to-add-syntax-highlight-to-sql-line-magic-cell-magic-and-custom-command-in) )

Is there any way to do that also in JupyterLab without developing an extension? 
Currently, the `custom.js` file is not recognized by JupyterLab.

Thank you

**Comment by bollwyvl** [comment]
Of note: we don't yet have the "good bits" (e.g. completion, linting, etc) for a SQL [language](https://github.com/joe-re/sql-language-server) [server](https://github.com/lighttiger2505/sqls) already connected, but these would not be a huge amount of effort: usually just one one config file, or a config file, an example, and a few tests if we intend to keep it tested and supported. Feel free to open an issue over there: even if it doesn't make it into the release, we can likely figure out something that works well enough.

**Comment by julioyildo** [comment]
Hi @krassowski ,

Thank you very much for your responses.
After having some troubles during the installation, we finally succeeded. 

We have tested the `%%sql` magic and that works fine! Great job!  Thank you! 

Is it possible to activate the same SQL recognition to other magics cell libraries? 
For example to[ IPython Magics for BigQuery  ](https://googleapis.dev/python/bigquery/latest/magics.html) : `%%bigquery`

Does the [Jupyter-lsp](https://github.com/krassowski/jupyterlab-lsp) extension gives the possibility to activate the SQL language server recognition (Highlight) on custom magic-cells?  We would like to add this highlighting to internal magic cells.

Thank you for your responses! :)

**Comment by krassowski** [comment]
> Is it possible to activate the same SQL recognition to other magics cell libraries?
> For example to IPython Magics for BigQuery : %%bigquery

Yes, we already do that for `%%R`, `%%javascript`, `%%latex` and dozens of others. We welcome a PR. It would be basically copy-pasting [src/transclusions/ipython-sql](https://github.com/krassowski/jupyterlab-lsp/tree/master/packages/jupyterlab-lsp/src/transclusions/ipython-sql) into `src/transclusions/ipython-bigquery` and adjusting the magic arguments patterns, and then adding it into the list of default transclusions [here](https://github.com/krassowski/jupyterlab-lsp/blob/master/packages/jupyterlab-lsp/src/transclusions/defaults.ts). The `language` and `file_extension` arguments would remain `'sql'` as BigQuery is a variant of SQL (so it can be highlighted as if it was SQL)*. The extractors protocol is documented [here](https://github.com/krassowski/jupyterlab-lsp/blob/a823c987b4c6d14dc755a6976e8a77f1f6bac774/packages/jupyterlab-lsp/src/extractors/regexp.ts#L111-L155).

> Does the Jupyter-lsp extension gives the possibility to activate the SQL language server recognition (Highlight) on custom magic-cells? We would like to add this highlighting to internal magic cells.

In principle yes. We are testing against [joe-re/sql-language-server](https://github.com/joe-re/sql-language-server) (and others listed [here](https://jupyterlab-lsp.readthedocs.io/en/latest/Language%20Servers.html)) but if you already installed it (which requires [a very specific version](https://github.com/joe-re/sql-language-server/issues/56) of node) and do not see highlights (but diagnostics/linting works) it might be that this server does not implement the highlights feature. I am regularly using `%%R` cells and highlights work great here so it should work for you too as long as the SQL language server provides highlights feature. There are other SQL language servers, some of which may be more recent and more advanced; we can assist you with connecting those to jupyterlab-lsp (and contributing what you learned along the way back to the community) but I personally would not have time to play around with different SQL language servers at the moment.

*) while a CodeMirror mode for BigQuery [was created](https://stackoverflow.com/a/23945333/6646912) many years ago by Google it seems to be an internal tool and I am not sure if it still works; if those are just tweaks to the `sql` highlighting mode that can be made using the public CodeMirror API we might consider how to enable adjusting mode parameters from transclusions API, but adding a new mode would be out of scope (especially because sql and bigquery are not that different).

**Comment by julioyildo** [comment]
Hi @krassowski ,

Thank you for your answers,

I followed the steps you described to be able to integrate `%%bigquery` as a magic `sql` cell.  Here is the link to my repo:
https://github.com/julioyildo/jupyterlab-lsp/tree/bigquery-magic 

However, after several attempts I encounter some installation problems. Indeed, I would like to be able to test `%%bigquery` in my local environment by installing the changes made on my repo.

I tried to install my changes in the following way: 

```
Pip install git+https://github.com/julioyildo/jupyterlab-lsp.git@bigquery-magic#egg=jupyterlab-lsp

Pip install python-language-server [all]
```

I have also followed the [contributing](https://github.com/krassowski/jupyterlab-lsp/blob/master/CONTRIBUTING.md) instructions but unsuccessful unfortunately...

Would you know how I can install and test my changes locally please? :)  (These are my first steps in the development of Jupyterlab extensions.)

Thank you ! :)

Note: 
My environment includes Nodejs 14 and Jupyterlab 3.0.8.

**Comment by krassowski** [comment]
No, installing the repository via pip will not work as it has multiple python and typescript packages. Please follow the instructions in contributing.md.

We can help more once you send a pull request with your contribution or open a reproducible bug report after following the instructions from contribution.md. Please open the PR and/or issue in the jupyterlab-lsp repository (not here) following the templates that will show up.

In general I would recommend to always first setup the environment, confirm that you can build the extension and only after that try to introduce changes because now after the changes are already introduced those might prevent compilation if you made any mistakes.

**Comment by kristoffSC** [comment]
Hi,
I managed to add syntax highlights to my custom magic similar like %%bigquery uses SQL. 
I rebuild lsp extension and jypiterlab accordingly to instructions from CONTRIBUTING.md. Everything seems to work properly.
Is this still the only way to add highlight support for custom magic? Doing this via config files is still on its way right?

On the other side I'm a bit confused since I don't have any Language server for SQL although the syntax highlights works. I assume I dont need Language Server if the only thing I need is highlights right?

Also I've noticed that "WITH" key word in SQL query is not highlighted. If I would like to modify this, add some "Key Words" that should be highlighted as well, where I should I do that?

**Comment by krassowski** [comment]
> Doing this via config files is still on its way right?

Yes, still on the roadmap - but if you would like to contribute to bring it faster, I will be happy to discuss in the jupyter-lsp repo.

> Also I've noticed that "WITH" key word in SQL query is not highlighted. If I would like to modify this, add some "Key Words" that should be highlighted as well, where I should I do that?

That might require modifying the CodeMirror mode - but I am not sure why WITH is not highlighted for you it seems to work on https://codemirror.net/mode/sql/ - do you have a reproducer?

**Comment by kristoffSC** [comment]
I think I'm to novice to contribute ;)

Regarding the "WITH". This is en example:
```
WITH totalSalary(Airline, total) as
    (SELECT Airline, sum(Salary)
    FROM Pilot
    GROUP BY Airline),
    airlineAverage(avgSalary) as 
    (SELECT avg(Salary)
    FROM Pilot )
    SELECT Airline
    FROM totalSalary, airlineAverage
    WHERE totalSalary.total > airlineAverage.avgSalary;
```
It does work on https://codemirror.net/mode/sql/  but on my jupiterlabI the "WITH" is not highlighted.
